Add automatic RPC configuration.

hectorj opened the pull request:
Implements #120 

When running btcd, if no configuration file is found, the sample-btcd.conf file is copied to the default path, with the rpcuser and rpcpass lines populated with random base64 encoded values.

When running btcctl, if no configuration file is found, it attempts to read the btcd.conf file at its default path, and create a basic btcctl.conf file by copying the rpcuser and rpcpass.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/btcsuite/btcd/609)
<!-- Reviewable:end -->

davecgh commented:
Thanks for the PR.  I haven't review this yet, but I noticed it failed the CI checks.  Looking at the Travis logs I see the following reasons:

> config.go:884:2: var generatedRpcUser should be generatedRPCUser
> config.go:890:2: var generatedRpcPass should be generatedRPCPass

hectorj commented:
I forgot to run goclean.sh, thanks.
Now fixed.

davecgh commented on cmd/btcctl/config.go:
```diff
@@ -268,3 +277,59 @@ func loadConfig() (*config, []string, error) {
+	}
+	defer dest.Close()
+
+	dest.WriteString(fmt.Sprintf(`[Application Options]
```
The `[Application Options]` header isn't required, so I would prefer it's left off.  Also, please do this using a regular string with \n so it doesn't mess the source code formatting up.

```Go
	dest.WriteString(fmt.Sprintf("rpcuser=%s\nrpcpass=%s",
		string(userSubmatches[1]), string(passSubmatches[1])))
```

hectorj replied:
Done (sorry for the delay)

hectorj commented:
Rebased on latest master and squashed

davecgh commented:
I just tested this out and the generated `conf` file contains extra new lines after every line:

```
; ------------------------------------------------------------------------------

; Data settings

; ------------------------------------------------------------------------------
```

davecgh commented:
`btcctl` is able to properly extract the `rpcuser` and `rpcpass` and generates the conf file automatically when it doesn't exist as expected.  That is super nice!

hectorj commented:
New lines issue fixed

davecgh commented on config.go:
```diff
@@ -880,6 +891,66 @@ func loadConfig() (*config, []string, error) {
+		}
+
+		if strings.Contains(line, "rpcuser=") {
+			line = "  rpcuser=" + string(generatedRPCUser) + "\n"
```
I don't believe this should have the extra 2 spaces at the front (nor the `rpcpass` below).

It ends up like the following:

```
; full set of credentials - limited or admin - or the RPC server will
; be disabled.
  rpcuser=<user>
  rpcpass=<pass>
; rpclimituser=whatever_limited_username_you_want
; rpclimitpass=
```

Instead of:

```
; full set of credentials - limited or admin - or the RPC server will
; be disabled.
rpcuser=<user>
rpcpass=<pass>
; rpclimituser=whatever_limited_username_you_want
; rpclimitpass=
```

davecgh commented:
Confirmed.  The only thing left is the extra two spaces before `rpcuser` and `rpcpass` that I commented on.  After that this it good to go.

hectorj commented:
- removed extra spaces
- removed debug code leftover in test (I was printing the generated file to check the newlines and forgot to remove it)
- refactored test by using ioutil.ReadFile instead of os.Open + ioutil.ReadAll
- squashed it all

davecgh commented:
Thanks.  I reviewed all the changes and tested it as follows:

- Ran btcd with existing conf file to ensure it didn't overwrite
- Ran btcctl with existing conf file to ensure it didn't overwrite
- Removed both btcd and btcctl conf files
- Ran btcd and confirmed it created the config file and formatted it appropriately
- Ran btcctl and confirmed it created the config file with the rpcuser and rpcpass extracted from the config file

Everything looks good.

OK

waxmihs2902 commented:
Reviewed 1 of 1 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/btcsuite/btcd/609)*
<!-- Sent from Reviewable.io -->
